### PR TITLE
Add Typesense Integration Tests To CI

### PR DESCRIPTION
**Pull request by sdreyer**

### CODE REVIEW

File: .github/workflows/integration-tests.yml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+      - name: Start Typesense
+        run: |
+          export TYPESENSE_API_KEY=xyz
+          mkdir /tmp/typesense-data
+          docker run -d -p 8108:8108 -v/tmp/typesense-data:/data typesense/typesense:0.22.2 \
+          --data-dir /data --api-key=$TYPESENSE_API_KEY --enable-cors
+      - name: Health Check
+        run: |
+          sleep 10
```
Comment (simba-git) [inline .github/workflows/integration-tests.yml:50]: Our typesense could should probably be setup with retry, could you add that

File: .github/workflows/integration-tests.yml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+      - name: Install Typesense Container
+        run: docker pull typesense/typesense:0.22.2
+
+      - name: Start Typesense
+        run: |
+          export TYPESENSE_API_KEY=xyz
+          mkdir /tmp/typesense-data
+          docker run -d -p 8108:8108 -v/tmp/typesense-data:/data typesense/typesense:0.22.2 \
+          --data-dir /data --api-key=$TYPESENSE_API_KEY --enable-cors
```
Comment (simba-git) [inline .github/workflows/integration-tests.yml:47]: I wonder if you can use the containers part of actions

**Review by simba-git — CHANGES_REQUESTED** [review]

File: .github/workflows/integration-tests.yml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+          git pull
+          git checkout ${{ github.base_ref }}
+          git pull
+          git merge ${{ github.head_ref }}
+      - name: Set up Go
+        uses: actions/setup-go@v2
+        with:
+          go-version: 1.17
+
+      - name: Install grpc_tools
```
[the remark this answers is not in the document]
Comment (Sami1309) [reply .github/workflows/integration-tests.yml:27]: Artifacts let you save data between jobs, eg.

name: Upload artifact
        uses: actions/upload-artifact@v2
        with:
          name: myimage
          path: /tmp/myimage.tar

**Review by Sami1309 — COMMENTED** [review]
comments don't pertain to core functionality, just minor possible improvements. Otherwise LGTM

File: .github/workflows/integration-tests.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (sdreyer) [reply .github/workflows/integration-tests.yml:27]: Good point. Ill create a setup job beforehand that initializes everything that all integration tests can use

File: ff/serving/metadata/search/search_test.go [hunk 4]
```diff
@@ -5,6 +5,9 @@ import (
 )
 
 func TestFullSearch(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (sdreyer) [reply ff/serving/metadata/search/search_test.go:7]: The testing package can take in a -short flag which triggers testing.Short(). The way I've set it up right now is that unit tests are run with go test -short, which skips integration tests. Running without the -short flag will run the integration tests

File: .github/workflows/integration-tests.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (sdreyer) [reply .github/workflows/integration-tests.yml:27]: Now builds /ff/serving/ and uploads the directory for other jobs to pull and use

File: .github/workflows/integration-tests.yml (hunk 2 above)
Comment (sdreyer) [reply .github/workflows/integration-tests.yml:63]: No easy way to pass in configuration args for the default typesense container using github actions at the moment. Would be good to use containers for other services though. Could create a custom container for typesense, but feels unnecessary atm

File: .github/workflows/integration-tests.yml (hunk 1 above)
Comment (sdreyer) [reply .github/workflows/integration-tests.yml]: Added retry into typesense calls

File: .github/workflows/integration-tests.yml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+          export TYPESENSE_API_KEY=xyz
+          mkdir /tmp/typesense-data
+          docker run -d -p 8108:8108 -v/tmp/typesense-data:/data typesense/typesense:0.22.2 \
+          --data-dir /data --api-key=$TYPESENSE_API_KEY --enable-cors
+      - name: Health Check
+        run: |
+          sleep 10
+          curl http://localhost:8108/health
+      - name: Integration Test
+        working-directory: ./ff/serving/metadata/search
```
[the remark this answers is not in the document]
Comment (sdreyer) [reply .github/workflows/integration-tests.yml]: Done

**Review by simba-git — APPROVED** [review]

**Review by Sami1309 — APPROVED** [review]
lgtm

**Pull request merged**